### PR TITLE
[go server] replace panics with returned errors

### PR DESCRIPTION
**Pull request by dmueller**
### problem
It is idiomatic in go to return errors rather than panic. Panics can cause applications to exit, which is not the desired result if a glean log can't be emitted for any reason.

### solution
Switch the exported functions to return an error. This enables applications that use the glean server implementation to inspect that error return value and take whatever action they deem appropriate.

A couple other things changed in the go server template
* made tabs consistently used in the file
* switched variable declarations to use `:=` syntax

### testing
`make test-full`
also added tests for `Writer = nil` and `Writer = io.Discard` to verify logs are not printed in those scenarios

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - `make test` runs without emitting any warnings
  - `make lint` runs without emitting any errors
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry to `CHANGELOG.md` or an explanation of why it does not need one
  - Any breaking changes to language binding APIs are noted explicitly

### CODE REVIEW

File: glean_parser/templates/go_server.jinja2 [hunk 1]
```diff
@@ -182,28 +195,28 @@ func newGleanEvent(category, name string, extra map[string]string) gleanEvent {
 {% for event in events %}
 
 type {{ event|event_type_name }} struct {
-    {% for extra, metadata in event.extra_keys.items() %}
-    {{ extra|event_extra_name }} {{ metadata.type|go_metric_type }} // {{ metadata.description|clean_string }}
-    {% endfor %}
+	{% for extra, metadata in event.extra_keys.items() %}
```
Comment (dmueller) [inline glean_parser/templates/go_server.jinja2:198]: This template was mixed tabs & spaces, so this whitespace change is to get everything on tabs

File: glean_parser/templates/go_server.jinja2 [hunk 2]
```diff
@@ -224,52 +237,52 @@ func (e {{ metric|event_type_name}}) is{{ ping|ping_events_type_name }}() {}
...
 }
 
 // Record and submit `{{ ping }}` ping omitting user request info
 func (g GleanEventsLogger) Record{{ ping|ping_type_name}}WithoutUserInfo(
-    params {{ ping|ping_type_name}},
-) {
-    g.Record{{ ping|ping_type_name }}(defaultRequestInfo, params)
+	params {{ ping|ping_type_name}},
+) error {
+	return g.Record{{ ping|ping_type_name }}(defaultRequestInfo, params)
```
Comment (dmueller) [inline glean_parser/templates/go_server.jinja2:286]: These two generated methods got their signature changed to return error, and the corresponding return statement

File: tests/test_go_server.py [hunk 3]
```diff
@@ -138,8 +138,9 @@ def run_logger(code_dir, code):
     with open(ROOT / "test-go" / "test.go.tmpl", "r") as fp:
         tmpl_code = fp.read()
 
-    tmpl_code = tmpl_code.replace("/* CODE */", code)
+    tmpl_code = tmpl_code.replace("/* CODE */", code).replace("/* IMPORTS */", imports)
 
+    print(tmpl_code)
```
Comment (cbguder) [inline tests/test_go_server.py:143]: Stray print?

**Review by cbguder — APPROVED** [review]

**Review by efixler — APPROVED** [review]
nice

**Review by chutten — APPROVED** [review]
I'm not a Go expert, but what I understand of these changes looks good to me. (The stray `print` should probably go, though. But it's test-only, so I'm not fussed if it makes it in)

File: tests/test_go_server.py (hunk 3 above)
Comment (dmueller) [reply tests/test_go_server.py:143]: TY, that was from getting the new tests working

**Review by akkomar — COMMENTED** [review]
+1 for error handling approach.

I hate to be starting this discussion (lol) but this repo generally uses spaces, not tabs, so these files will have them both mixed eventually. I know Go's recommendation is to use tabs, but IIUC `gofmt` should address that. Is this change useful on your side when working with generated file or is it mostly about consistency across files here?

**Comment by akkomar** [comment]
> Is there a reason to make the jinja templates all use spaces?

Per https://github.com/mozilla/glean_parser/blob/main/.editorconfig#L5-L6 space is default indentation character for these files.

**Comment by dmueller** [comment]
I didn't know about the editorconfig code and am using vscode without a plugin for it, so that's good to know about!

**Review by akkomar — APPROVED** [review]
